src/cuda/cuda_topk.cu

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cub/device/device_segmented_radix_sort.cuh>
+
+#include "cuda_topk.h"
+#include "cuda_topk_full_sort.cuh"
+#include "cuda_topk_select_sort.cuh"
+
+namespace Generators {
+namespace cuda {
+
+TopkData::TopkData(int batch_size, int vocab_size, cudaStream_t stream) {
+  size_t vocab_batch_size = static_cast<size_t>(vocab_size) * batch_size;
+
+  // Allocate all necessary device memory
+  // The buffers are sized for the full sort algorithm, which is the largest.
+  intermediate_indices_1 = CudaMallocArray<int>(vocab_batch_size);
+  intermediate_indices_2 = CudaMallocArray<int>(vocab_batch_size);
+  intermediate_scores_1 = CudaMallocArray<float>(vocab_batch_size);
+  intermediate_scores_2 = CudaMallocArray<float>(vocab_batch_size);
+  batch_offsets = CudaMallocArray<int>(batch_size + 1);
+
+  cub_temp_storage_bytes = GetFullSortCubTempStorageBytes(vocab_batch_size, batch_size, stream);
+  cub_temp_storage = CudaMallocArray<unsigned char>(this->cub_temp_storage_bytes);
+}
+
+// Kernel to compact strided data into a dense layout.
+// Used to convert data from a [batch, stride] layout to a dense [batch, k] layout.
+template <typename T>
+__global__ void CompactStridedData(const T* input, T* output, int k, int batch_size, int input_stride) {
+  const int batch_idx = blockIdx.x;
+  for (int i = threadIdx.x; i < k; i += blockDim.x) {
+    int in_idx = batch_idx * input_stride + i;
+    int out_idx = batch_idx * k + i;
+    output[out_idx] = input[in_idx];
+  }
+}
+
+void TopkDataCompact::CompactOutput(int batch_size, int vocab_size, cudaStream_t stream, int k) {
+  topk_scores_compact = CudaMallocArray<float>(static_cast<size_t>(batch_size) * k);
+  topk_indices_compact = CudaMallocArray<int>(static_cast<size_t>(batch_size) * k);
+  dim3 grid(batch_size);
+  dim3 block(256);
+  CompactStridedData<float><<<grid, block, 0, stream>>>(topk_scores, topk_scores_compact.get(), k, batch_size, topk_stride);
+  CompactStridedData<int><<<grid, block, 0, stream>>>(topk_indices, topk_indices_compact.get(), k, batch_size, topk_stride);
+  cudaStreamSynchronize(stream);
+}
+
+void GetTopK(TopkData* topk_data, cudaStream_t stream, const float* scores_in, int vocab_size, int batch_size, int k) {
+  assert(topk_data != nullptr);
+
+  if (k > 64) {
+    LaunchSort(topk_data, stream, scores_in, topk_data->intermediate_scores_1.get(),
+               topk_data->intermediate_indices_1.get(), vocab_size, batch_size);
+  } else {
+    // NOTE: This modifies scores_in in-place
```

Comment:
Maybe expose a boolean flag to indicate input buffer mutation is allowed and default it to true ? That way future users of this routine will be aware from the parameter list that potentially the input buffer could be modified if they provide `true` for that flag